### ISSUE TITLE
Support returning functions in `match` patterns

### ISSUE THREAD

**Issue by MaaaxiKing** [godotengine/godot-proposals#2337]
### Describe the project you are working on
Reaction game
### Describe the problem or limitation you are having in your project
I cannot check for a text with `tr()` in `match`.
### Describe the feature / enhancement and how it helps to overcome the problem or limitation
You could use any function that returns something as a pattern in a match statement.
### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
```
match $Label.text:
	tr("foo"):
		print("hi")
	tr("bar"):
		print("ha")
```
### If this enhancement will not be used often, can it be worked around with a few lines of script?
No
### Is there a reason why this should be core and not an add-on in the asset library?
Yes, it's useful for everyone and not even possible to implement with an asset afaik.

**Comment by Calinou** [comment]
Try this instead:

```gdscript
match tr($Label.text):
	"foo":
		print("hi")
	"bar":
		print("ha")
```

**Comment by MaaaxiKing** [comment]
I tried and it did not work which also surprised me.

**Comment by MaaaxiKing** [comment]
What you wrote makes that if the Label's text is "MY TEXT.", it will check if there's a translation with the key "MY TEXT." instead of checking for the translation.

**Comment by Calinou** [comment]
@MaaaxiKing Out of curiosity, what's the end goal here? What are you trying to achieve?

**Comment by MaaaxiKing** [comment]
I have a tutorial and there's text to display. Everytime the user does what he has to, I call `next()` and there I want to check what the current text ist. My current way is that I have a variable: `index`. And in `next()` I check for this with `match` and at the end of the function I increase `index` by 1.

**Comment by Calinou** [comment]
> I have a tutorial and there's text to display. Everytime the user does what he has to, I call `next()` and there I want to check what the current text ist. My current way is that I have a variable: `index`. And in `next()` I check for this with `match` and at the end of the function I increase `index` by 1.

I think you should have a separate variable that stores the untranslated text key at all times and use that for the `match` comparison instead.

**Comment by cameronsteele** [comment]
I agree that this would be very useful; at least from a code-aesthetic standpoint. It doesn't make much sense to me for Godot to take such a firm stance on this aspect of the architecture. Now, if I want to compare a value against function calls, I am forced to use if/else simply because Godot's syntax is so opinionated.

@Calinou Unfortunately, the solution you provided above would not generally work, because `tr($Label.text)` would return a different value depending on the language the client was using; so it would only actually work with languages where the return was "foo" or whatever.

In a use case for this feature (specific to OP's and, by chance, my own), we likely have something like this for a File Menu:
```
var fileOptions = [
    tr("NEW"),
    tr("SAVE"),
    tr("SAVE_AS"),
    tr("OPEN")
]

for fileOption in fileOptions:
    fileMenuPopup.add_item(fileOption)
```

and then later in a menu select callback:

```
...
	match menu:
		"file":
			match fileOptions[id]:
				tr("NEW"):
					addBlankLevel()

				tr("SAVE"):
					saveDialog.popup()

				tr("SAVE_AS"):
					saveAsDialog.popup()

				tr("OPEN"):
					loadDialog.popup()
```

This doesn't seem unreasonable to me. There's an argument to be made that mapping the localization keys to constants could be a clever idea if syntax allows it; but I don't think that keeping function calls out of match patterns makes much sense to me. It ought to be left up to the programmer to decide whether that is a point in the code where they want to optimize, no?

**Comment by DaGamingWolf** [comment]
i fully support this. it seems incredibly weird not to be able to use anything in a match statement that isn't a constant. it has hugely limited my code. i keep running into this over and over again. can't use variables as a pattern? this is silly. you should easily be able to compare a string with a bunch of different node names that vary because you have no idea what those names will be. if if statements can do it, there's no reason a match statement shouldn't also be able to.

**Comment by DaGamingWolf** [comment]
maybe this is the reason variables can no longer be used in match statements in 4.0. setters and getters return the value. the loss of the ability to use variables in match statements is way to huge a loss imo not to reconfigure match to allow returning values for comparison. it still seems incredibly weird to me that this is an issue at all.

**Comment by dalexeev** [comment]
* Resolved by godotengine/godot#72285.
